server/auth/types/openid/routes.ts

```diff
@@ -0,0 +1,213 @@
+        if (!request.query.code) {
+          const nonce = randomString(22);
+          const query: any = {
+            client_id: this.config.openid?.client_id,
+            response_type: 'code',
+            redirect_uri: `${this.getBaseRedirectUrl()}`, // TODO: generate actual domain name
+            state: nonce,
+          };
+
+          const scope = this.config.openid?.scope || '';
+          if (scope) {
+            query.scope = scope;
+          }
+
+          const queryString = stringify(query);
+          const location = `${this.openIdAuthConfig.authorizationEndpoint}?${queryString}`;
+          const cookie: SecuritySessionCookie = {
+            oidcState: nonce,
+          };
+          this.sessionStorageFactory.asScoped(request).set(cookie);
+          return response.redirected({
+            headers: {
+              location: location,
+            },
+          });
+        }
+
+        // Authentication callback
+        // TODO: figure out a decent way to validate the state from cookie
+        const clientId = this.config.openid?.client_id;
+        const clientSecret = this.config.openid?.client_secret;
+        const query: any = {
+          grant_type: 'authorization_code',
+          code: request.query.code,
+          redirect_uri: `${this.getBaseRedirectUrl()}`,
+          client_id: clientId,
+          client_secret: clientSecret,
+        };
+
+        const requestOptions: any = {
+          payload: stringify(query),
+          headers: {
+            'Content-Type': 'application/x-www-form-urlencoded',
+          },
+        };
+
+        try {
+          const tokenResponse = await wreck.post(
+            this.openIdAuthConfig.tokenEndpoint as string,
+            requestOptions
+          );
+          if (
+            !tokenResponse.res.statusCode ||
+            tokenResponse.res.statusCode < 200 ||
+            tokenResponse.res.statusCode > 299
+          ) {
+            return response.unauthorized();
+          }
+
+          const tokenPayload: any = this.parseTokenResponse(tokenResponse.payload as Buffer);
+
+          const idToken: string = tokenPayload.id_token;
+          const accessToken: string = tokenPayload.access_token;
+          const refreshToken: string = tokenPayload.refresh_token;
+          const expiresIn: number = tokenPayload.expires_in;
+
+          const user = await this.securityClient.authenticateWithHeader(
+            request,
+            this.openIdAuthConfig.authHeaderName as string,
+            `Bearer ${idToken}`
+          );
+
+          // set to cookie
+          const sessionStorage: SecuritySessionCookie = {
+            username: user.username,
+            credentials: {
+              authHeaderValue: `Bearer ${idToken}`,
+              refresh_token: refreshToken,
+              expires_at: Date.now() + expiresIn * 1000,
+            },
+            authType: 'openid',
+            expiryTime: Date.now() + this.config.cookie.ttl,
```

Comment:
Minor: probably include unit into the ttl variable to make it easy to set the value without worrying about the unit.

Reply:
this is a config in kibana.yml, want to keep the parity with the existing implementation